diff --git a/drivers/ufs/core/ufshcd.c b/drivers/ufs/core/ufshcd.c
--- a/drivers/ufs/core/ufshcd.c
+++ b/drivers/ufs/core/ufshcd.c
@@ -2112,7 +2112,7 @@ static void ufshcd_init_clk_gating(struct ufs_hba *hba)
 
 	ufshcd_init_clk_gating_sysfs(hba);
 
-	hba->clk_gating.is_enabled = true;
+	hba->clk_gating.is_enabled = false;
 	hba->clk_gating.is_initialized = true;
 }
 
@@ -10550,12 +10550,8 @@ int ufshcd_init(struct ufs_hba *hba, void __iomem *mmio_base, unsigned int irq)
 	/*
 	 * Set the default power management level for runtime and system PM.
 	 * Default power saving mode is to keep UFS link in Hibern8 state
 	 * and UFS device in sleep state.
 	 */
-	hba->rpm_lvl = ufs_get_desired_pm_lvl_for_dev_link_state(
-						UFS_SLEEP_PWR_MODE,
-						UIC_LINK_HIBERN8_STATE);
-	hba->spm_lvl = ufs_get_desired_pm_lvl_for_dev_link_state(
-						UFS_SLEEP_PWR_MODE,
-						UIC_LINK_HIBERN8_STATE);
+	hba->rpm_lvl = UFS_PM_LVL_0;
+	hba->spm_lvl = UFS_PM_LVL_0;
 
 	INIT_DELAYED_WORK(&hba->rpm_dev_flush_recheck_work,
 			  ufshcd_rpm_dev_flush_recheck_work);
